sdk/src/main/java/com/amplitude/experiment/DefaultExperimentClient.kt

```diff
@@ -105,20 +111,44 @@ internal class DefaultExperimentClient internal constructor(
 
     override fun variant(key: String, fallback: Variant?): Variant {
         val variantAndSource = resolveVariantAndSource(key, fallback)
-        val variant = variantAndSource.variant
-        val source = variantAndSource.source
         if (config.automaticExposureTracking) {
-            exposureInternal(key, variant, source)
+            exposureInternal(key, variantAndSource)
         }
-        return variant
+        return variantAndSource.variant
     }
 
     override fun exposure(key: String) {
-        val variantAndSource = resolveVariantAndSource(key, null)
-        exposureInternal(key, variantAndSource.variant, variantAndSource.source)
+        val variantAndSource = resolveVariantAndSource(key)
+        exposureInternal(key, variantAndSource)
+    }
+
+    private fun exposureInternal(key: String, variantAndSource: VariantAndSource) {
+        legacyExposureInternal(key, variantAndSource.variant, variantAndSource.source)
+
+        // Do not track exposure for fallback variants that are not associated with a default variant.
+        val fallback = isFallback(variantAndSource.source)
+        if (fallback && (variantAndSource.hasDefaultVariant == null || !variantAndSource.hasDefaultVariant)) {
+            return
+        }
+
+        val experimentKey = variantAndSource.variant.expKey
+        var variant: String? = null
+        val metadata = variantAndSource.variant.metadata
+        if (!fallback && (metadata?.get("default") == null)) {
+            variantAndSource.variant.key?.let {
+                variant = it
+            } ?: variantAndSource.variant.value?.let {
+                variant = it
+            }
```

Comment:
Nit: the syntax using `let` to check if the value is null including a chained elvis operator is very confusing.
```
variant = variantAndSource.variant.key ?: variantAndSource.variant.value
```

IMO this is a lot more clear, then you can actually make the if statement return a value and turn `variant` into an immutable value:

```
val variant = if (...) {
  variantAndSource.variant.key ?: variantAndSource.variant.value
}
```

Reply:
The reason we do all this finnegaling with javascript is to reduce the size of the Javascript object when it is serialized. We dont have this problem in other non dynamic languages.